lib/spack/spack/solver/concretize.lp

```diff
@@ -51,10 +50,38 @@ depends_on(Package, Dependency, Type)
   depends_on(Package, Provider, Type)
   : provides_virtual(Provider, Virtual)
 } 1
- :- declared_dependency(Package, Virtual, Type),
+ :- dependency_conditions(Package, Virtual, Type),
     virtual(Virtual),
-    not external(Package),
-    node(Package).
+    not external(Package).
+
+% if any individual condition below is true, trigger the dependency.
+dependency_conditions(P, D, T) :- dependency_conditions(P, D, T, _).
+
+% collect all the dependency condtions into a single conditional rule
+dependency_conditions(P, D, T, I) :-
+  node(Package)
+    : dep_cond(P, D, T, I, "node", Package);
+  version(Package, Version)
+    : dep_cond(P, D, T, I, "version", Package, Version);
+  version_satisfies(Package, Constraint)
+    : dep_cond(P, D, T, I, "version_satisfies", Package, Constraint);
+  node_platform(Package, Platform)
+    : dep_cond(P, D, T, I, "node_platform", Package, Platform);
+  node_os(Package, OS)
+    : dep_cond(P, D, T, I, "node_os", Package, OS);
+  node_target(Package, Target)
+    : dep_cond(P, D, T, I, "node_target", Package, Target);
+  variant_value(Package, Variant, Value)
+    : dep_cond(P, D, T, I, "variant_value", Package, Variant, Value);
+  node_compiler(Package, Compiler)
+    : dep_cond(P, D, T, I, "node_compiler", Package, Compiler);
+  node_compiler_version(Package, Compiler, Version)
+    : dep_cond(P, D, T, I, "node_compiler_version", Package, Compiler, Version);
+  node_flag(Package, FlagType, Flag)
+    : dep_cond(P, D, T, I, "node_flag", Package, FlagType, Flag);
+  dependency_condition(P, D, I);
+  declared_dependency(P, D, T);
+  node(P).
```

Comment:
Here I think there's redundancy that can be simplified. In a term like:
```
dep_cond(P, D, T, I, "node", Package)
```
I don't think we'll ever encounter a case where `P != Package`. Possibly we may have something like:
```
node(Package) : required_condition(Package, ID, "node");
version(Package, Version) : required_condition(Package, ID, "version", Version);
```
Since having the `Package` and the `ID` should uniquely identify a `depends_on` directive in a repository.

Reply:
>I don't think we'll ever encounter a case where P != Package.

My reasoning above has a flaw, as it doesn't account for statements that involve dependencies in a condition:
```python
depends_on('patchelf@0.9', when='@1.0:1.1 ^python@:2')
```
In #20426 I gave a global ID to `depends_on` and `conflicts` directives to disambiguate among them. The resulting ASP condition looks like:
```
  node(Package)
    : required_dependency_condition(ID, "node", Package);
  version(Package, Version)
    : required_dependency_condition(ID, "version", Package, Version);
```
For a `depends_on` similar to the above the `node` condition will ground to:
```
node(this_package), node(python), ...
```
due to the use of the `^` sigil in the `when` argument of `depends_on`.

Reply:
>>I don't think we'll ever encounter a case where P != Package.
>
>My reasoning above has a flaw, as it doesn't account for statements that involve dependencies in a condition:
>```python
>depends_on('patchelf@0.9', when='@1.0:1.1 ^python@:2')
>```

Yep -- this is why I used `P` there instead of `Package` :).  But I see this is fixed in your later commit, which is now fast-forwarded here.